File: src/routes/documents.js
const express = require('express');
const router = express.Router();
const db = require('../models/db');
const { v4: uuidv4 } = require('uuid');

// GET docs for a repository
router.get('/repository/:repositoryId', async (req, res) => {
  try {
    const docs = await db.documents.findAsync({ repositoryId: req.params.repositoryId });
    docs.sort((a, b) => new Date(b.updatedAt) - new Date(a.updatedAt));
    res.json({ success: true, data: docs });
  } catch (err) {
    res.status(500).json({ success: false, error: err.message });
  }
});

// GET single document
router.get('/:id', async (req, res) => {
  try {
    const doc = await db.documents.findOneAsync({ _id: req.params.id });
    if (!doc) return res.status(404).json({ success: false, error: 'Document not found' });
    res.json({ success: true, data: doc });
  } catch (err) {
    res.status(500).json({ success: false, error: err.message });
  }
});

// POST create document
router.post('/', async (req, res) => {
  try {
    const {
      repositoryId,
      type,      // 'overview' | 'code' | 'example' | 'changelog' | 'algorithm' | 'guide'
      title,
      content,
      metadata   // type-specific structured data
    } = req.body;

    if (!repositoryId || !title?.trim()) {
      return res.status(400).json({ success: false, error: 'repositoryId and title required' });
    }

    const repo = await db.repositories.findOneAsync({ _id: repositoryId });
    if (!repo) return res.status(404).json({ success: false, error: 'Repository not found' });

    const doc = await db.documents.insertAsync({
      _id: uuidv4(),
      repositoryId,
      type: type || 'overview',
      title: title.trim(),
      content: content || '',
      metadata: metadata || {},
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    });

    res.json({ success: true, data: doc });
  } catch (err) {
    res.status(500).json({ success: false, error: err.message });
  }
});

// PUT update document
router.put('/:id', async (req, res) => {
  try {
    const { title, content, metadata, type } = req.body;
    await db.documents.updateAsync(
      { _id: req.params.id },
      { $set: { title, content, metadata, type, updatedAt: new Date().toISOString() } }
    );
    const updated = await db.documents.findOneAsync({ _id: req.params.id });
    res.json({ success: true, data: updated });
  } catch (err) {
    res.status(500).json({ success: false, error: err.message });
  }
});

// DELETE document
router.delete('/:id', async (req, res) => {
  try {
    await db.documents.removeAsync({ _id: req.params.id }, {});
    res.json({ success: true });
  } catch (err) {
    res.status(500).json({ success: false, error: err.message });
  }
});

module.exports = router;

File: src/routes/documents.py
"""
src/routes/documents.py
CRUD endpoints for Documents.
"""
import uuid
import json
from datetime import datetime, timezone
from bottle import request, response
from src.models.database import get_conn


def now_iso():
    return datetime.now(timezone.utc).isoformat()


def register(app):

    @app.get('/api/documents/repository/<repo_id>')
    def list_docs(repo_id):
        with get_conn() as conn:
            rows = conn.execute(
                """SELECT * FROM documents WHERE repository_id = ?
                   ORDER BY created_at ASC""",
                (repo_id,)
            ).fetchall()
        return {'success': True, 'data': [dict(r) for r in rows]}

    @app.get('/api/documents/<doc_id>')
    def get_doc(doc_id):
        with get_conn() as conn:
            row = conn.execute(
                'SELECT * FROM documents WHERE id = ?', (doc_id,)
            ).fetchone()
        if not row:
            response.status = 404
            return {'success': False, 'error': 'Document not found'}
        return {'success': True, 'data': dict(row)}

    @app.post('/api/documents')
    def create_doc():
        body = request.json or {}
        title = (body.get('title') or '').strip()
        repo_id = (body.get('repository_id') or '').strip()
        if not title or not repo_id:
            response.status = 400
            return {'success': False, 'error': 'title and repository_id are required'}
        doc_id = str(uuid.uuid4())
        ts = now_iso()
        metadata = body.get('metadata', {})
        if isinstance(metadata, dict):
            metadata = json.dumps(metadata)
        with get_conn() as conn:
            conn.execute(
                """INSERT INTO documents
                   (id, repository_id, doc_type, title, content, metadata, created_at, updated_at)
                   VALUES (?, ?, ?, ?, ?, ?, ?, ?)""",
                (doc_id, repo_id,
                 body.get('doc_type', 'overview'),
                 title,
                 body.get('content', ''),
                 metadata, ts, ts)
            )
            row = conn.execute('SELECT * FROM documents WHERE id = ?', (doc_id,)).fetchone()
        return {'success': True, 'data': dict(row)}

    @app.put('/api/documents/<doc_id>')
    def update_doc(doc_id):
        body = request.json or {}
        title = (body.get('title') or '').strip()
        if not title:
            response.status = 400
            return {'success': False, 'error': 'title is required'}
        ts = now_iso()
        metadata = body.get('metadata', {})
        if isinstance(metadata, dict):
            metadata = json.dumps(metadata)
        with get_conn() as conn:
            conn.execute(
                """UPDATE documents
                   SET doc_type=?, title=?, content=?, metadata=?, updated_at=?
                   WHERE id=?""",
                (body.get('doc_type', 'overview'),
                 title,
                 body.get('content', ''),
                 metadata, ts, doc_id)
            )
            row = conn.execute('SELECT * FROM documents WHERE id = ?', (doc_id,)).fetchone()
        if not row:
            response.status = 404
            return {'success': False, 'error': 'Document not found'}
        return {'success': True, 'data': dict(row)}

    @app.delete('/api/documents/<doc_id>')
    def delete_doc(doc_id):
        with get_conn() as conn:
            conn.execute('DELETE FROM documents WHERE id = ?', (doc_id,))
        return {'success': True}

    @app.get('/api/documents/search')
    def search_docs():
        query = (request.query.get('q') or '').strip().lower()
        if not query:
            return {'success': True, 'data': []}
        pattern = f'%{query}%'
        with get_conn() as conn:
            rows = conn.execute(
                """SELECT * FROM documents
                   WHERE lower(title) LIKE ? OR lower(content) LIKE ?
                   ORDER BY updated_at DESC LIMIT 20""",
                (pattern, pattern)
            ).fetchall()
        return {'success': True, 'data': [dict(r) for r in rows]}
